### ISSUE TITLE
fleet rebalancing feature

### ISSUE THREAD

**Issue by rimusz** [coreos/fleet#1023]
As per our discussion on https://groups.google.com/forum/#!topic/coreos-user/2TMksA4tMfQ

Currently fleet nicely balances units around cluster machines only for:
1) new fresh installed cluster 
2) adding new units to the current cluster

The problem is if one of node gets reboot e.g. an update or we get a new node in the cluster, units do not get rebalanced back around all the nodes.

I would suggest to have a simple rebalancing  for the fleet.
1) Rebooted/new node pings fleet - I'm here give me some job :) 
2) fleet calculates all units which can be rebalanced, counts them, checks cluster nodes number and 
randomly picks units and rebalances them. (the same procedure you use for deploying new units should be fine there)

e.g.
3 x nodes cluster with 20 units which can work on any node.
1) 1 node gets rebooted for the upgrade.
2) 20/3=6.66 round it to 7
3) random 7 units get rebalanced on the rebooted/new node.

CPU, RAM usage and etc should not be counted in, we do not want to complicated things too much :)

**Comment by lynchc** [comment]
+1

Any progress on this?

**Comment by sevos** [comment]
+1?

**Comment by nballotta** [comment]
+1

**Comment by tjanczuk** [comment]
For people who have a similar scenario to mine, here is how I worked around this limitation of fleet:

1. I created a template unit file `my_service@.service` with `Global=true` setting
2. I spawn X instances of this unit in the cluster, say 10. Given a 3 VM cluster, this results in 30 units running across the cluster. 
3. In case of a single VM failure, the VM just takes the instances that were running on it down with it. Fleet *does not* reschedule them on other VMs - but this is something I can live with given my requirements. 
4. When a new VM joins or re-joins the cluster, the `Global=true` setting in the unit template file will automatically have 10 instances of my_service started on that VM.

**Comment by lynchc** [comment]
I talked to some people on the irc channel about it. It makes sense to me now to keep it as a simple/dumb scheduler. if you need anything that handles redistribution better, it makes better sense to jump into something like kubernetes

**Comment by zebulonj** [comment]
My two cents:  Given that a fundamental feature of coreos is automatic restarts for upgrades, and the fact that these restarts redistribute units, there has to be some minimal way of redistributing after restarts complete.

Example:
Say I'm running two nodes in the cluster designated (via metadata) as targets for x@.service.  Initially, I load x@1 and x@2, and these units are appropriately distributed so that one is on each of the two nodes. Subsequently, an update is distributed and node-1 restarts (say that locking ensures that only one node at a time is down). As a result, fleet restarts x@1 on node-2. When node-1 comes back up, it sits idle. Then node-2 gets restarted and both x@1 and x@2 are rescheduled for node-1.  When it comes back up, node-2 is idle.

It seems that it will always be the case that without some intervention (either human or a scheduling layer on top of fleet), the cluster will only ever use N-1 nodes after an update is distributed.  If an extra layer of scheduling is required on top of fleet to deal with this, how can fleet really be called a scheduler?

FWIW, the work around above from @tjanczuk appears to be the best option short of another layer. Doesn't seem like a work-around should be necessary for this simple, ubiquitous use case.

**Comment by brianwhigham** [comment]
+1  Sadly, I think I'll have to abandon fleet before this issue is resolved.

**Comment by tylerFowler** [comment]
+1 on this if it's being considered for a feature. I wouldn't necessarily want anything very in depth, as others have mentioned that's what heavier duty things are for.

My use case is very simple, say I have a specific set of worker pool servers and I want to run 6 worker containers total (assuming here they depend on nothing, and if they do then all the pool servers have it). When deploying they are evenly split across the servers, when CoreOS performs a reboot all of them end up on the servers that didn't reboot and stay that way. All I would need in this scenario is a periodic check to ensure that the nodes are correctly balanced, and if not then proactively rebalancing them. Perhaps even a lock on fleet's etcd keys from locksmithd when a reboot has been scheduled to prevent fleet from thinking the server is actually down (though that may not be viable on larger deployments).

**Comment by joeatwork** [comment]
Could this be the kind of thing that could be handled by an external client of fleet rather than needing built-in support? That way questions like "should fleet ignore scheduled CoreOS reboots" could be handled site by site?

**Comment by tylerFowler** [comment]
I agree, in my particular case the issue is directly caused by a feature of CoreOS (reboots) so it may in the realm of some service like locksmith.

**Comment by Gibby** [comment]
I have been looking for something like this for awhile but would like to see this included in fleet so an option could be added to the .service files under X-Fleet. Something like allow_auto_rebalance=true.
I have tried other options including Kubernetes, Mesosphere, Panamax, Rancher. However they either can't do it, require more containers therefor more resources or are way to complicated for something this simple.


1 Clients use case:
2 Physical hosts
4 CoreOS vm's, 2 on each physical host
1 vm on each machine is just for quorum checking and has low memory assigned but higher CPU priority than the other VM's.
The other vm's are the ones that do the workloads. 

The workload vm's have enough resources to handle running all the containers for a few hours until some of the periodic workloads kick off.

We have implemented a crude script to check if the number of containers is balanced between worker nodes and if not, do some simple checks(uptime of rebooted node, input file for containers allowed to be moved at anytime and a couple other things) and then rebalance them. While this is getting them by in the interim, it would be great to see this supported in fleet and if not in fleet, maybe in a helper container/service.

**Comment by yuyumao** [comment]
+1

**Comment by jairovm** [comment]
It'd be really nice to have this 👍
